### ISSUE TITLE
AttributeError: module 'boxsdk.object' has no attribute 'collaboration'

### ISSUE THREAD

**Issue by c-rei** [box/box-python-sdk#368]
Our applications are working very well with boxsdk version 1.5.5.
When we update boxsdk to version 2.0..0.0, our application can work at source. But if we freeze our sources into stand-alone executables with Pyinstaller, we got an error like below.
 
Traceback (most recent call last):
  File "Box-Report\boxreport.py", line 9, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "c:\users\changqli\appdata\local\anaconda3\envs\pydev36\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
    exec(bytecode, module.__dict__)
  File "site-packages\boxsdk\__init__.py", line 8, in <module>
AttributeError: module 'boxsdk.object' has no attribute 'collaboration'
[4548] Failed to execute script boxreport

We want to solve this issue, can we get any help? 
Our environments are windows 7, 64 bits and Pyinstaller version 3.4.

**Comment by jmoldow** [comment]
Although we know that we sometimes have problems with freezers (see #162), it is unclear to me how this particular module could have regressed. `boxsdk.object.collaboration` exists in both 1.5.5 and 2.0.0, and is in `boxsdk.object.__all__` in both of those versions.

The only thing I see that is different is that `__all__` was a single line of code in 1.5.5, and is now multiple lines of code. Maybe that somehow has confused Pyinstaller.

The quickest thing to do to fix this would be for you to statically import everything you know you need, somewhere inside your project. That way the freezer will not optimize those modules out of the build.

For example:

```python
# Inside one of your files. Add in any other types that you might encounter in your program
from boxsdk.object import (collaboration, file, folder, enterprise, user, web_link)
```

**Comment by eliasIBM** [comment]
Hi, is this issue going to be resolved?

**Comment by eliasIBM** [comment]
To anyone that is struggling with this as well, you need to edit the hidden imports attribute in the spec file to get all of the modules. For me it was just the modules in the boxsdk.object folder.

`# -*- mode: python -*-

block_cipher = None


a = Analysis(['FileViewerApp.py'],
             pathex=['/Users/eliasyishak/Desktop/USPS/USPS-PVS/Front End Tool/Box Environment'],
             binaries=[],
             datas=[],
             hiddenimports=[
                'boxsdk.object.collaboration',
                'boxsdk.object.collaboration_whitelist',
                'boxsdk.object.collaboration_whitelist_entry',
                'boxsdk.object.collaboration_whitelist_exempt_target',
                'boxsdk.object.collection',
                'boxsdk.object.comment',
                'boxsdk.object.device_pinner',
                'boxsdk.object.enterprise',
                'boxsdk.object.events',
                'boxsdk.object.event',
                'boxsdk.object.file',
                'boxsdk.object.file_version',
                'boxsdk.object.file_version_retention',
                'boxsdk.object.folder',
                'boxsdk.object.group',
                'boxsdk.object.group_membership',
                'boxsdk.object.legal_hold',
                'boxsdk.object.legal_hold_policy',
                'boxsdk.object.legal_hold_policy_assignment',
                'boxsdk.object.metadata_cascade_policy',
                'boxsdk.object.metadata_template',
                'boxsdk.object.recent_item',
                'boxsdk.object.retention_policy',
                'boxsdk.object.retention_policy_assignment',
                'boxsdk.object.search',
                'boxsdk.object.storage_policy',
                'boxsdk.object.storage_policy_assignment',
                'boxsdk.object.terms_of_service',
                'boxsdk.object.terms_of_service_user_status',
                'boxsdk.object.trash',
                'boxsdk.object.task',
                'boxsdk.object.task_assignment',
                'boxsdk.object.user',
                'boxsdk.object.upload_session',
                'boxsdk.object.webhook',
                'boxsdk.object.watermark',
                'boxsdk.object.web_link'],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          [],
          name='FileViewerApp',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          runtime_tmpdir=None,
          console=False)
`

**Comment by c-rei** [comment]
Hey @eliasIBM 

Thanks your comments. 

I wander why I must to import all modules of boxsdk.object expressly, don't just imported boxsdk.object to my spec file like this.

hiddenimports=['boxsdk.object']

**Comment by eliasIBM** [comment]
I’m not 100% sure why, it must have to do with how the different modules are being imported within boxsdk.objects and pyinstaller. PyInstaller isn’t smart enough to realize how these modules are included so you must tell it explicitly where to go look and find the modules you need included in the packaged executable.

**Comment by Sivadutta** [comment]
> Hi Sivadutta,
> 
> I'm not following what you mean by distributing the spec file. The spec file is generated if you run pyinstaller with a python script as the argument. But you can also run the spec file in place of the python script after it has been generated. The command below is an example of how to run it from your terminal/command prompt:
> 
> pyinstaller mygeneratedspecfile.spec
> 
> Hope this helps.

Hi, I got it after raising the question. However, the issues are never ending it seems. Now I am dealing with: 
1. I have used a json config file(generated from box), placed it in the working directory & reading it in py file. Now this json file need to go with exe generated through pyinstaller. The challenge is include it in spec file correctly & generate exe. Unable to do this so far...still googling. 
Another question: If succeeded to generate exe, do I need to distribute the exe along with the json config file or not!
2. As an workaround of json config file, copy pasted all values into the py file & used it. Not sure, whether it is a good idea or not.

Thanks for your help. I am just 1 month old to Python, Box & all open source related techs.

Sivadutta

**Comment by eliasIBM** [comment]
Hi,

I have had this issue before and I’m sure there are ways to import a json file but I’ve had better luck just converting the json file into a py file and importing that file as an object into my main script.

That should work unless you are going to be editing the json/py file after the exe has been created.

**Comment by Sivadutta** [comment]
> Hi,
> 
> I have had this issue before and I’m sure there are ways to import a json file but I’ve had better luck just converting the json file into a py file and importing that file as an object into my main script.
> 
> That should work unless you are going to be editing the json/py file after the exe has been created.

Hi,

Thanks for replying. I did the same after searching countless suggestions. It is working fine now. I faced another issue while running the exe in another system but I think I am able to run it now.
1. Exe generated through pyinstaller, copied the exe to a Azure window running on Win server 2016.
2. Created a scheduler, created a batch file(to execute the python exe). Attach the scheduler to run the batch file. Got error here, it never executed.
3. I have a logger file writer also in my python file.

Again googled whole day about the execution of python exe on a server which doesn't have python installed in it. If I double click the bat file, then everything works fine. So, I decided to put following lines in the bat file:
cmd.exe /c 
cd C:\<Scriptfolder name> 
"<Scriptfolder name>\exename.exe"

After lot of try... finally the scheduler is invoking the batch file & the python exe is doing it's job. I will put it every where to help some people. Everyday I am learning new things.... :)

Thanks,
Sivadutta
